src/uast.cc

```diff
@@ -137,95 +244,155 @@ void *UastIteratorNext(UastIterator *iter) {
-  int nodeIdx = 0;
-  xmlNodeSetPtr result = nullptr;
-  xmlNodePtr *results = nullptr;
-  xmlGenericErrorFunc handler;
-
-  Nodes *nodes;
-  try {
-    nodes = new Nodes();
-  } catch (std::bad_alloc) {
-    Error(nullptr, "Unable to get memory\n");
+  auto result = UastFilterTyped(ctx, node, query);
+
+  if (!checkResult(result, XPATHTYPE_NODESET, "UastFilter")) {
     return nullptr;
   }
-  doc = CreateDocument(ctx, node);
-  if (!doc) {
-    goto error1;
-  }
-  xpathCtx = static_cast<xmlXPathContextPtr>(xmlXPathNewContext(doc));
-  if (!xpathCtx) {
-    goto error2;
-  }
 
-  handler = (xmlGenericErrorFunc)Error;
-  initGenericErrorDefaultFunc(&handler);
+  return result.nodesVal;
+}
+
+int UastFilterBool(const Uast *ctx, void *node, const char *query) {
+  auto result = UastFilterTyped(ctx, node, query);
 
-  xpathObj = xmlXPathEvalExpression(BAD_CAST(query), xpathCtx);
-  if (!xpathObj) {
-    goto error3;
+  if (!checkResult(result, XPATHTYPE_BOOLEAN, "UastFilterBoolean")) {
+    return -1;
   }
 
-  // Get array of results
-  result = xpathObj->nodesetval;
-  if (!result) {
-    Error(nullptr, "Unable to get array of results\n");
-    goto error3;
-  }
-  results = result->nodeTab;
-  size = (result) ? result->nodeNr : 0;
-
-  // Sometimes libxml return null results for some invalid queries; do
-  // a first pass to discard them and get the real size for the result
-  for (int i = 0; i < size; i++) {
-    if (results[i] != nullptr && results[i]->_private != nullptr) {
-      ++realSize;
-    }
-  }
+  return result.boolVal;
+}
 
-  if (NodesSetSize(nodes, realSize) != 0) {
-    Error(nullptr, "Unable to set nodes size\n");
-    goto error3;
-  }
+double UastFilterNumber(const Uast *ctx, void *node, const char *query) {
+  auto result = UastFilterTyped(ctx, node, query);
 
-  // Populate array of results
-  for (int i = 0; i < size; i++) {
-    if (results[i] != nullptr && results[i]->_private != nullptr) {
-      nodes->results[nodeIdx++] = results[i]->_private;
-    }
+  if (!checkResult(result, XPATHTYPE_NUMBER, "UastFilterNumber")) {
+    return -1;
   }
-  ok = true;
 
-error3:
-  xmlXPathFreeObject(xpathObj);
-error2:
-  xmlXPathFreeContext(xpathCtx);
-error1:
-  xmlFreeDoc(doc);
+  return result.numberVal;
+}
+
+const char *UastFilterString(const Uast *ctx, void *node, const char *query) {
+  auto result = UastFilterTyped(ctx, node, query);
 
-  if (!ok) {
-    NodesFree(nodes);
+  if (!checkResult(result, XPATHTYPE_STRING, "UastFilterString")) {
     return nullptr;
   }
-  return nodes;
+
+  return result.stringVal;
 }
 
-char *LastError(void) {
-  return strndup(error_message, BUF_SIZE);
+UastTypedResult UastFilterTyped(const Uast *ctx, void *node, const char *query) {
+  UastTypedResult ret;
+
+  try {
+
+    auto handler = static_cast<xmlGenericErrorFunc>(Error);
+    initGenericErrorDefaultFunc(&handler);
+
+    QueryResult qobject(ctx, node, query);
+    ret.type = static_cast<XPathType>(qobject.xpathObj->type);
```

Comment:
Maybe abstracting this could be a good idea (and just call `result->type()` here).
Probably all the result handling could be moved to `QueryResult` class, I don't know.

Reply:
Will go away with the refactor of this function in the next push of this PR.